Automattic/jetpack
# API Nonces: performance optimization and refactoring.

This is a revised version of the PR #17030, which was reverted due to unanticipated side effects.
In this PR nonce cleanup still relies on a WP-cron scheduled task, but the cleanup process is optimized for better performance.

The whole nonce code was moved out of the `Connection\Manager` and into a dedicated class `Nonce_Handler` for better code clarity and easier maintenance.

#### Changes proposed in this Pull Request:
- Extracting nonces code into a dedicated class `Nonce_Handler` for better clarity.
- Covering the functionality with unit tests.
- Optimizing the WP-cron scheduled query to make it run faster.
- Setting a forced 5-second limit on the scheduled cleanup so it wouldn't run for too long.

#### Jetpack product discussion
p9dueE-1Oi-p2
p9dueE-20a-p2

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
1. Connect Jetpack, install and activate the "WP Crontrol" plugin.
2. Use the Jetpack Debug's "Mocker" tool to create a bunch of old nonces (a few dozens would be enough, unless you want to test the performance too).
3. Run the MySQL query and confirm that the nonces were created:
```sql
SELECT COUNT(*) FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%';
```
4. Go to "Tools -> Cron Events" and run the `jetpack_clean_nonces` hook.
5. Fetch the date/time of the oldest nonce:
```sql
SELECT FROM_UNIXTIME(option_value, '%Y-%m-%d %H:%i:%s') FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%' ORDER BY option_value LIMIT 1;
```
6. Convert the UTC date to your local timezone, and make sure that all nonces older than one hour were removed.
7. Disconnect Jetpack. Confirm that all existing nonces got removed, including the fresh ones:
```sql
SELECT COUNT(*) FROM wp_options WHERE option_name LIKE 'jetpack\_nonce\_%';
```

#### Proposed changelog entry for your changes:
* Improving the Jetpack Nonce cleanup process.
